build-tools/packages/build-cli/src/library/packageExports.ts

```diff
@@ -0,0 +1,140 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+
+import type { Logger, PackageJson } from "@fluidframework/build-tools";
+
+import { ApiTag } from "./apiTag";
+
+export interface ExportData {
+	/**
+	 * Location of file relative to package
+	 */
+	relPath: string;
+	/**
+	 * Conditions required for export access; in hierarchy order
+	 */
+	conditions: string[];
+	/**
+	 * Export is only .d.ts file
+	 */
+	isTypeOnly: boolean;
+}
+export type Node10CompatExportData = Omit<ExportData, "conditions">;
+
+/**
+ * Only the value types of exports that are records.
+ */
+type ExportsRecordValue = Exclude<Extract<PackageJson["exports"], object>, unknown[]>;
+
+function findTypesPathMatching(
+	mapQueryPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined>,
+	exports: ExportsRecordValue,
+	conditions: string[],
+): (ExportData & { apiTagLevel: ApiTag | undefined }) | undefined {
+	for (const [entry, value] of Object.entries(exports)) {
+		if (typeof value === "string") {
+			if (entry === "types") {
+				for (const [key, apiTagLevel] of mapQueryPathToApiTagLevel.entries()) {
+					// eslint-disable-next-line max-depth
+					if (
+						typeof key === "string"
+							? path.resolve(value) === path.resolve(key)
+							: key.test(value)
+					) {
+						const isTypeOnly = !(
+							"default" in exports ||
+							"import" in exports ||
+							"require" in exports
+						);
+						return { apiTagLevel, relPath: value, conditions, isTypeOnly };
+					}
+				}
+			}
+		} else if (value !== null) {
+			if (Array.isArray(value)) {
+				continue;
+			}
+			const deepFind = findTypesPathMatching(mapQueryPathToApiTagLevel, value, [
+				...conditions,
+				entry,
+			]);
+			if (deepFind !== undefined) {
+				return deepFind;
+			}
+		}
+	}
+
+	return undefined;
+}
+
+export function queryOutputMapsFromPackageExports(
```

Comment:
Docs? Also the `mapQueryPathToApiTagLevel` param in particular could use some docs explaining the expected key format / what it means for the value to be undefined (with a key presents).